external/kafka-0-10-sql/src/main/scala/org/apache/spark/sql/kafka010/KafkaSource.scala

```diff
@@ -0,0 +1,446 @@
+        implicitly[Ordering[(String, Long)]].compare(
+          (l.topic, l.partition),
+          (r.topic, r.partition))
+      }
+    }
+    val sortedTopicPartitions = untilPartitionOffsets.keySet.toSeq.sorted(topicPartitionOrdering)
+    val sortedExecutors = getSortedExecutorList(sc)
+    val numExecutors = sortedExecutors.size
+    logDebug("Sorted executors: " + sortedExecutors.mkString(", "))
+    val offsetRanges = sortedTopicPartitions.flatMap { tp =>
+      fromPartitionOffsets.get(tp).map { fromOffset =>
+        val untilOffset = untilPartitionOffsets(tp)
+        val preferredLoc = if (numExecutors > 0) {
+          Some(sortedExecutors(positiveMod(tp.hashCode, numExecutors)))
+        } else None
+        KafkaSourceRDD.OffsetRange(tp, fromOffset, untilOffset, preferredLoc)
+      }
+    }.toArray
+
+    // Create a RDD that reads from Kafka and get the (key, value) pair as byte arrays.
+    val rdd = new KafkaSourceRDD[Array[Byte], Array[Byte]](
+      sc, executorKafkaParams, offsetRanges, sourceOptions).map { cr =>
+        Row(cr.checksum, cr.key, cr.offset, cr.partition, cr.serializedKeySize,
+          cr.serializedValueSize, cr.timestamp, cr.timestampType.id, cr.topic, cr.value)
+    }
+
+    logInfo("GetBatch: " + offsetRanges.sortBy(_.topicPartition.toString).mkString(", "))
+    sqlContext.createDataFrame(rdd, schema)
+  }
+
+  /** Stop this source and free any resources it has allocated. */
+  override def stop(): Unit = synchronized {
+    consumer.close()
+  }
+
+  override def toString(): String = s"KafkaSource[$consumerStrategy]"
+
+  private def fetchPartitionOffsets(seekToLatest: Boolean): Map[TopicPartition, Long] = {
+    synchronized {
+      logTrace("\tPolling")
+      consumer.poll(0)
+      val partitions = consumer.assignment()
+      consumer.pause(partitions)
+      logDebug(s"\tPartitioned assigned to consumer: $partitions")
+      if (seekToLatest) {
+        consumer.seekToEnd(partitions)
+        logDebug("\tSeeked to the end")
+      }
+      logTrace("Getting positions")
+      val partitionToOffsets = partitions.asScala.map(p => p -> consumer.position(p))
+      logDebug(s"Got positions $partitionToOffsets")
+      partitionToOffsets.toMap
+    }
+  }
+
+  private def positiveMod(a: Long, b: Int): Int = ((a % b).toInt + b) % b
+}
+
+/** Companion object for the [[KafkaSource]]. */
+private[kafka010] object KafkaSource {
+
+  def kafkaSchema: StructType = StructType(Seq(
+    StructField("checksum", LongType),
+    StructField("key", BinaryType),
+    StructField("offset", LongType),
+    StructField("partition", IntegerType),
+    StructField("serializedKeySize", IntegerType),
+    StructField("serializedValueSize", IntegerType),
+    StructField("timestamp", LongType),
+    StructField("timestampType", IntegerType),
+    StructField("topic", StringType),
+    StructField("value", BinaryType)
+  ))
+
+  sealed trait ConsumerStrategy[K, V] {
```

Comment:
The underlying Consumer api is half baked (well, to be fair, more like 85% baked).  We already had this exact same argument, it's already exposed in the DStream implementation.  If you try to encapsulate it, you're pretty much ensuring that the spark implementation has to change when kafka changes.  If you simply expose what kafka provides, the spark implementation isn't going to have to change much, if at all.